generator/metaline_builder.cpp

```diff
@@ -0,0 +1,217 @@
+    {
+      m_parts.push_back(line);
+      return;
+    }
+    for (LineString & part : m_parts)
+    {
+      if (part.Add(*found))
+      {
+        m_parts.erase(found);
+        break;
+      }
+    }
+  }
+
+  std::vector<Ways> GetLongWays() const
+  {
+    std::vector<Ways> result;
+    for (LineString const & line : m_parts)
+      if (line.Ways().size() > 1)
+        result.push_back(line.Ways());
+    return result;
+  }
+};
+
+void MetalinesBuilder::operator()(OsmElement const & el, FeatureParams const & params)
+{
+  static uint32_t const highwayType = classif().GetTypeByPath({"highway"});
+  if (params.FindType(highwayType, 1) == ftype::GetEmptyValue() ||
+      el.Nodes().front() == el.Nodes().back())
+    return;
+
+  std::string name;
+  params.name.GetString(0, name);
+  if (name.empty() || params.ref.empty())
+    return;
+
+  std::size_t const key = std::hash<std::string>{}(name + '\0' + params.ref);
+  auto segment = m_data.find(key);
+  if (segment == m_data.cend())
+    m_data.emplace(key, std::make_shared<Segments>(el));
+  else
+    segment->second->Add(el);
+}
+
+void MetalinesBuilder::Flush()
+{
+  uint32_t count = 0;
+  FileWriter writer(m_filePath);
+  for (auto const & seg : m_data)
+  {
+    auto const & longWays = seg.second->GetLongWays();
+    for (Ways const & ways : longWays)
+    {
+      uint16_t size = ways.size();
+      WriteToSink(writer, size);
+      for (std::int32_t const way : ways)
+        WriteToSink(writer, way);
+      ++count;
+    }
+  }
+  LOG_SHORT(LINFO, ("Wrote", count, "metalines with OSM IDs for the entire planet to", m_filePath));
+}
+
+bool WriteMetalinesSection(std::string const & mwmPath, std::string const & metalinesPath,
+                           std::string const & osmIdsToFeatureIdsPath)
+{
+  map<osm::Id, uint32_t> osmIdToFeatureId;
+  if (!routing::ParseOsmIdToFeatureIdMapping(osmIdsToFeatureIdsPath, osmIdToFeatureId))
+    return false;
+
+  FileReader reader(metalinesPath);
+  ReaderSource<FileReader> src(reader);
+  std::vector<std::uint8_t> buffer;
+  MemWriter<std::vector<std::uint8_t>> memWriter(buffer);
+  std::uint32_t count = 0;
+
+  while (src.Size())
+  {
+    std::vector<std::int32_t> featureIds;
+    std::uint16_t size = ReadPrimitiveFromSource<std::uint16_t>(src);
+    std::vector<std::int32_t> ways(size);
+    src.Read(&ways[0], size * sizeof(std::int32_t));
+    for (auto const wayId : ways)
+    {
+      auto fid = osmIdToFeatureId.find(osm::Id::Way(std::abs(wayId)));
+      if (fid == osmIdToFeatureId.cend())
+        break;
+
+      featureIds.push_back(wayId > 0 ? fid->second : -fid->second);
+    }
+
+    if (featureIds.size() > 1)
+    {
+      // Write vector to buffer if we have got at least two segments.
+      WriteVarUint(memWriter, featureIds.size());
+      for (auto const & fid : featureIds)
+        WriteVarInt(memWriter, fid);
+      ++count;
+    }
+  }
+
+  // Terminate the metalines stream with a zero-length list.
+  uint8_t const zero = 0;
```

Comment:
As a way to signal the end of metalines stream. I don't know if the reader can catch the end of the mwm section, so I added this zero byte just in case.